core/src/consensus/tendermint/engine.rs

```diff
@@ -263,3 +290,267 @@ impl ConsensusEngine for Tendermint {
+        *proposed += 1;
+        // FIXME: Consider banned accounts
+        *missed += authors.len() - TendermintSealView::new(&header.seal()).bitset()?.count();
+    }
+
+    let mut reduced_rewards = 0;
+
+    // Penalty disloyal validators
+    let number_of_blocks_in_term = start_of_the_next_term - start_of_the_current_term;
+    for (address, intermediate_reward) in rewards {
+        // FIXME: Consider banned accounts
+        let number_of_signatures = u64::try_from(*signed_blocks.get(&address).unwrap()).unwrap();
+        let final_block_rewards = final_rewards(intermediate_reward, number_of_signatures, number_of_blocks_in_term);
+        reduced_rewards += intermediate_reward - final_block_rewards;
+        pending_rewards.insert(address, final_block_rewards);
+    }
+
+    // Give additional rewards
+    give_additional_rewards(reduced_rewards, missed_signatures, |address, reward| {
+        pending_rewards.insert(*address, reward);
+        Ok(())
+    })?;
+
+    Ok(pending_rewards)
+}
+
+/// reward = floor(intermediate_rewards * (a * number_of_signatures / number_of_blocks_in_term + b) / 10)
+fn final_rewards(intermediate_reward: u64, number_of_signatures: u64, number_of_blocks_in_term: u64) -> u64 {
+    let (a, b) = if number_of_signatures * 3 >= number_of_blocks_in_term * 2 {
+        // number_of_signatures / number_of_blocks_in_term >= 2 / 3
+        // x * 3/10 + 7/10
+        (3, 7)
+    } else if number_of_signatures * 3 >= number_of_blocks_in_term {
+        // number_of_signatures / number_of_blocks_in_term >= 1 / 3
+        // x * 24/10 - 7/10
+        (24, -7)
+    } else {
+        // 1 / 3 > number_of_signatures / number_of_blocks_in_term
+        // x * 3/10 + 0
+        assert!(
+            number_of_blocks_in_term > 3 * number_of_signatures,
+            "number_of_signatures / number_of_blocks_in_term = {}",
+            (number_of_signatures as f64) / (number_of_blocks_in_term as f64)
+        );
+        (3, 0)
+    };
+    let numerator = i128::from(intermediate_reward)
+        * (a * i128::from(number_of_signatures) + b * i128::from(number_of_blocks_in_term));
+    assert!(numerator >= 0);
+    let denominator = 10 * i128::from(number_of_blocks_in_term);
+    // Rust's division rounds towards zero.
+    u64::try_from(numerator / denominator).unwrap()
+}
+
+fn give_additional_rewards<F: FnMut(&Address, u64) -> Result<(), Error>>(
+    mut reduced_rewards: u64,
+    missed_signatures: HashMap<Address, (usize, usize)>,
+    mut f: F,
+) -> Result<(), Error> {
+    let lcm_of_proposed = missed_signatures
+        .values()
+        .map(|(proposed, _)| *proposed)
+        .collect::<HashSet<_>>()
+        .into_iter()
+        .fold(1usize, |base, proposed| base.checked_mul(proposed).unwrap());
+    let sorted_validators = missed_signatures
+        .into_iter()
+        .map(|(address, (proposed, missed))| (address, missed.checked_mul(lcm_of_proposed / proposed).unwrap()))
```

Comment:
I understand that `missed / proposed` cannot be calculated precisely without the least common multiplier. However, since we are comparing the non-negative ratio, we can compare `(missed/proposed, missed%proposed)` to eliminate `lcm_of_proposed`

Reply:
LGTM. I'll change it

Reply:
I changed it to use `num-rational`.